RxSwift/Observables/Implementations/ForwardIf.swift

```diff
@@ -0,0 +1,129 @@
+//
+//  ForwardIf.swift
+//  Rx
+//
+//  Created by Jorge Bernal Ordovas on 20/01/16.
+//  Copyright © 2016 Krunoslav Zaher. All rights reserved.
+//
+
+import Foundation
+
+class ForwardIfSinkIter<S: ObservableType, O: ObserverType where O.E == S.E> : ObserverType {
+    typealias Parent = ForwardIfSink<S, O>
+    typealias DisposeKey = CompositeDisposable.DisposeKey
+    typealias E = O.E
+
+    private let _parent: Parent
+    private let _disposeKey: DisposeKey
+
+    init(parent: Parent, disposeKey: DisposeKey) {
+        _parent = parent
+        _disposeKey = disposeKey
+    }
+
+    func on(event: Event<E>) {
+        switch event {
+        case .Next(let value):
+            _parent._lock.lock(); defer { _parent._lock.unlock() } // lock {
+                _parent.forwardOn(.Next(value))
+            // }
+        case .Error(let error):
+            _parent._lock.lock(); defer { _parent._lock.unlock() } // lock {
+                _parent.forwardOn(.Error(error))
+                _parent.dispose()
+            // }
+        case .Completed:
+            _parent._group.removeDisposable(_disposeKey)
+            _parent._lock.lock(); defer { _parent._lock.unlock() } // lock {
+                _parent.forwardOn(.Completed)
+                _parent.dispose()
+            // }
+        }
+    }
+
+}
+
+class ForwardIfSink<S: ObservableType, O: ObserverType where O.E == S.E>: Sink<O>, ObserverType {
+    typealias E = Bool
+    typealias SourceType = S.E
+    typealias DisposeKey = CompositeDisposable.DisposeKey
+
+    private let _lock = NSRecursiveLock()
+
+    private let _source: S
+    private let _group = CompositeDisposable()
+    private let _conditionSubscription = SingleAssignmentDisposable()
+    private var _sourceDisposeKey: DisposeKey? = nil
+
+    init(observer: O, source: S) {
+        _source = source
+        super.init(observer: observer)
+    }
+
+    func on(event: Event<E>) {
+        switch event {
+        case .Next(let element):
+            if element {
+                subscribeInner(_source)
```

Comment:
You would also need to synchronize calling `subscribeInner` and `unsubscribeInner` since they are mutating self.